### PR TITLE
feat(presets): add rdg2D preset for 2D-virtualized react-data-grid

### LINKED ISSUE #306: feat: add presets.rdg2D for 2D-virtualized react-data-grid tables
## What

Add a `presets.rdg2D` preset for react-data-grid tables where **both rows and columns are virtualized simultaneously** (2D virtualization). The existing `presets.rdg` handles 1D (column-only) virtualization but does not include the `viewport` strategy needed to safely read cells on 2D grids.

## Why

On 2D virtualized grids, PST's `_navigateToCell` holds a row locator reference while scrolling horizontally to reach off-screen columns. Because horizontal scrolling triggers vertical recycling in RDG, the row is removed from the DOM mid-navigation and PST exhausts its retry budget:

```
Error: SmartTable: could not reach cell for column "id" (colIndex 0) at row 37
after exhausting navigation strategies.
```

The `viewport` strategy already solves this — it detects when a row has been knocked out of the vertical viewport by a horizontal scroll and calls `scrollToRow` to restore it before reading. The types even document this exact use case and include an `aria-rowindex / aria-colindex` DOM fallback example. It just isn't wired into a preset.

## Proposed preset

```ts
export const rdg2D = {
  ...rdg, // inherits rowSelector, headerSelector, getCellLocator, dedupe
  strategies: {
    ...rdg.strategies,
    viewport: {
      getVisibleRowRange: async ({ root }) =>
        root.evaluate(el => {
          const rows = [...el.querySelectorAll('[role="row"][aria-rowindex]')];
          const idxs = rows.map(r => Number(r.getAttribute('aria-rowindex')) - 1);
          return { first: Math.min(...idxs), last: Math.max(...idxs) };
        }),

      getVisibleColumnRange: async ({ root }) =>
        root.evaluate(el => {
          const cells = [...el.querySelectorAll('[role="gridcell"][aria-colindex]')];
          const idxs = cells.map(c => Number(c.getAttribute('aria-colindex')) - 1);
          return idxs.length
            ? { first: Math.min(...idxs), last: Math.max(...idxs) }
            : { first: 0, last: 0 };
        }),

      scrollToRow: async ({ root }, rowIndex) =>
        root.evaluate((el, idx) => {
          const row = el.querySelector(`[role="row"][aria-rowindex="${idx + 1}"]`);
          if (row) {
            const top = parseInt(row.style.top || '0', 10);
            el.scrollTop = Math.max(0, top - el.clientHeight / 3);
          } else {
            // Row not in DOM — estimate using typical RDG row height
            el.scrollTop = Math.max(0, idx * 35 - el.clientHeight / 3);
          }
        }, rowIndex),

      scrollToColumn: async ({ root }, colIndex) =>
        root.evaluate((el, idx) => {
          const header = el.querySelector(`[role="columnheader"][aria-colindex="${idx + 1}"]`);
          if (header) {
            el.scrollLeft = header.offsetLeft;
          } else {
            const headers = [...el.querySelectorAll('[role="columnheader"][aria-colindex]')];
            if (headers.length) {
              const avg = headers.reduce((s, h) => s + h.getBoundingClientRect().width, 0) / headers.length;
              el.scrollLeft = idx * avg;
            }
          }
        }, colIndex),
    },
  },
};
```

## Notes

- The 35px row height fallback in `scrollToRow` is an estimate. A `rowHeight` option could be added to the preset for grids with non-standard row heights.
- For grids where column headers are 2D-virtualized (not just rows), a custom `header` strategy is also needed. The preset could document this as an extension point.
- `presets.rdg2D` was validated against a Grafana Explore react-data-grid with 13 columns and 50 rows, all fully virtualized on both axes. `row.toJSON()` and `table.map()` both worked correctly after adding the viewport strategy.

### PR DESCRIPTION
**Pull request by rickcedwhat-ai**
Closes #306

## Summary

- Adds `presets.rdg2D` — a drop-in replacement for `presets.rdg` for react-data-grid tables where both rows **and** columns are virtualized simultaneously
- Implements a custom `ViewportStrategy` with four RDG-specific methods that prevent "could not reach cell" errors caused by horizontal scrolling evicting rows from the DOM
- Exports `RDG2DStrategies` for callers who want strategies only (same pattern as `RDGStrategies`)

## Viewport strategy details

| Method | Behaviour |
|---|---|
| `getVisibleRowRange` | Queries `[role="row"][aria-rowindex]`, converts 1-based RDG indices to 0-based |
| `getVisibleColumnRange` | Queries `[role="gridcell"][aria-colindex]`, converts 1-based to 0-based |
| `scrollToRow` | Reads `style.top` when the row is in DOM; falls back to **measuring height of a visible row** (not hardcoded 35px) when the row has been evicted |
| `scrollToColumn` | Jumps via `aria-colindex` header `offsetLeft`; falls back to average visible column width |

## Test plan

- [ ] New `tests/integration/rdg-2d.spec.ts` — 4 tests against `comcast.github.io/react-data-grid`:
  - Headers collected (≥15 including virtualized columns)
  - `toJSON()` reads all columns without errors
  - `map()` collects 50+ unique rows across pages with no stale-locator errors
  - `getCell` works across different horizontal column positions

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Introduced the new `rdg2D` preset, enabling 2D virtualization for efficient handling of large data grids with optimized row and column scrolling capabilities.

* **Tests**
  * Added integration tests to verify 2D virtualization functionality.

### CODE REVIEW

**Comment by rickcedwhat-ai** [comment]
## 🤖 Bot HQ


### 🔗 Issue Link
🔗 Closes #306


---




### 🔍 AI Review
- [ ] 🔍 Request AI review

> ⏳ Review requested for #313 — AI review in progress…


---
<sub>This comment is managed by the bot — do not edit directly.</sub>

**Comment by rickcedwhat-ai** [comment]
## AI Review

PR #313 adds the rdg2D preset for React Data Grid with 2D virtualization. This is a well-implemented feature with proper TypeScript typing, comprehensive tests, and thoughtful handling of DOM eviction edge cases.
> ✅ No actionable issues found.

**Review by rickcedwhat — APPROVED** [review]

File: tests/integration/rdg-2d.spec.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        const dataRows = rows.filter((r: any) => r.ID !== 'Total');
+        expect(dataRows.length).toBeGreaterThanOrEqual(50);
+
+        // All IDs must be valid positive integers
+        expect(dataRows.every((r: any) => /^\d+$/.test(String(r.ID)))).toBe(true);
+
+        // No duplicates
+        const uniqueIds = new Set(dataRows.map((r: any) => r.ID));
+        expect(uniqueIds.size).toBe(dataRows.length);
```
[the remark this answers is not in the document]
Comment (rickcedwhat-ai) [reply tests/integration/rdg-2d.spec.ts]: Fixed in 515f998 — replaced the three inline `any` casts with a local `type RdgRow = { ID: string; Task?: string }` and a single cast on the `map` result, so `filter`, `every`, and `map` callbacks are all typed.

File: tests/integration/rdg-2d.spec.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        const grid = page.locator('[role="grid"]').first();
+        await expect(grid).toBeAttached({ timeout: 10000 });
+
+        const table = useTable(grid, { ...presets.rdg2D });
+        await table.init();
+
+        const rows = await table.findRows({}, { maxPages: 1 });
+        const firstRow = rows[0];
+
+        const data = await firstRow.toJSON({ columns: ['ID', 'Task', 'Completion'] });
```
[the remark this answers is not in the document]
Comment (rickcedwhat-ai) [reply tests/integration/rdg-2d.spec.ts]: Fixed in 515f998 — added `expect(rows.length).toBeGreaterThan(0)` immediately after `findRows` and inlined `rows[0]` directly, removing the `firstRow` intermediate variable.

**Comment by rickcedwhat-ai** [comment]
## AI Review — Round 2

This PR adds the `rdg2D` preset for react-data-grid with 2D virtualization support. The implementation is well-structured and properly integrates a viewport strategy to handle horizontal and vertical scrolling. The test coverage is comprehensive and the code follows existing patterns. No critical issues found.
> ✅ No actionable issues found.
